File: solutions/ruby/dnd-character/1/dnd_character.rb
=begin
Write your code for the 'D&D Character' exercise in this file. Make the tests in
`dnd_character_test.rb` pass.

To get started with TDD, see the `README.md` file in your
`ruby/dnd-character` directory.
=end

class DndCharacter
  attr_reader :strength, :dexterity, :constitution, :intelligence, :wisdom, :charisma, :hitpoints
  
  BASE_HITPOINTS = 10
  
  def self.modifier(c)
    (c - BASE_HITPOINTS).div(2)
  end

  def initialize
    @strength, @dexterity, @constitution, @intelligence, @wisdom, @charisma = Array.new(6) { new_ability }
    @hitpoints = BASE_HITPOINTS + self.class.modifier(@constitution)
  end

  private
    def new_throw
      Array.new(4) { Random.rand(1..6) }
    end

    def new_ability
      new_throw.sort.drop(1).sum
    end
end

File: solutions/clojure/dnd-character/1/src/dnd_character.clj
(ns dnd-character)


(defn score-modifier
  "Calculates the modifier of the given score."
;; You find your character's constitution modifier by subtracting 10 from your character's constitution, divide by 2 and round down.
  [score]
  (Math/floorDiv (- score 10) 2))

(defn- throw-die
  "Generates a random ability by throwing a 6 side dice"
  []
  (inc (rand-int 6)))

(defn rand-ability []
  (->> (repeatedly throw-die)
       (take 4)
       sort
       rest
       (apply +)))

(def abilities [:strength :dexterity :constitution :intelligence :wisdom :charisma])

(defn- add-hit-points [m]
  (->> m
       :constitution
       score-modifier
       (+ 10)
       (assoc m :hitpoints)))

(defn rand-character
  "Generates a random character."
  []
  (->> (repeatedly rand-ability)
       (take 6)
       (zipmap abilities)
       (add-hit-points)))
